src/wp-includes/media.php

```diff
@@ -987,9 +984,111 @@ function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $icon
+	return apply_filters( 'wp_get_attachment_preview_src', $image, $attachment_id, $size );
+}
+
+/**
+ * Retrieves an image to represent an attachment.
+ *
+ * @since 2.5.0
+ *
+ * @param int          		$attachment_id Image attachment ID.
+ * @param string|int[] 		$size          Optional. Image size. Accepts any registered image size name, or an array of
+ *                                    	   width and height values in pixels (in that order). Default 'thumbnail'.
+ * @param bool|string|array $args          {
+ *     Optional. An array of arguments to determine the image retrieved. Default empty array.
+ *
+ *     @type boolean $icon      Whether to return the mime type icon if an image is not available
+ *     @type string  $mime_type Define the attachment mime type to return. Must be a valid mime type (e.g ‘image/webp’).
+ *                              Will return the requested mime type if available or the attachment's original mime type
+ *                              as a fallback.
+ * }
+ * @return array|false {
+ *     Array of image data, or boolean false if no image is available.
+ *
+ *     @type string $0 Image source URL.
+ *     @type int    $1 Image width in pixels.
+ *     @type int    $2 Image height in pixels.
+ *     @type bool   $3 Whether the image is a resized image.
+ * }
+ */
+function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $args = [] ) {
+	// Determine the image mime type to return, falling back to the original mime type.
+	$target_mime   = null;
+	$original_mime = get_post_mime_type( $attachment_id );
+
+	// Determine the correct mime type to use.
+	if ( isset( $args['mime_type'] ) || is_string( $args ) ) {
+		$target_mime = isset( $args['mime_type'] ) ? $args['mime_type'] : $args;
+	}
+
+	if ( ! $target_mime ) {
+		$target_mime = $original_mime;
+	}
+
+	// Determine whether to fallback to an icon.
+	$icon = false;
+
+	if ( isset( $args['icon'] ) || is_bool( $args ) ) {
+		$icon = isset( $args['icon'] ) ? $args['icon'] : $args;
+	}
+
+	// Get the initial image.
+	$image = image_downsize( $attachment_id, $size );
+
+	// Get the correct mime if alternative is requested.
+	if ( $image && ( $target_mime !== $original_mime ) ) {
+		$metadata = wp_get_attachment_metadata( $attachment_id );
+
+		// Determine correct image size if size array passed.
+		if ( is_array( $size ) ) {
+			foreach( $metadata['sizes'] as $image_size => $size_data ) {
+				if ( $size_data['file'] !== $src[0] ) {
+					continue;
+				}
+				$size = $image_size;
+			}
+		}
+
+		if ( 'full' === $size && isset( $metadata['sources'][ $target_mime ]['file'] ) ) {
+			$file = wp_basename( $metadata['sources'][ $target_mime ]['file'] );
+		} elseif ( isset( $metadata['sizes'][ $size ]['sources'][ $target_mime ]['file'] ) ) {
+			$file = $metadata['sizes'][ $size ]['sources'][ $target_mime ]['file'];
+		}
+
+		$image[0] = str_replace( wp_basename( $image[0] ), $file, $image[0] );
+	}
```

Comment:
Looking at this code block, it seems inefficient to first get the image URL for the right sized image file and then replace it with the version for the correct MIME type. There's an extra lookup loop, several if checks and a string replacement.

I think it would make more sense to add an optional `$mime_type = ''` parameter to `image_downsize()`, and also to `image_get_intermediate_size()` which is called from there. This way we can really handle the MIME type selection at the lowest level where it's happening rather than first selecting the wrong MIME type and then fixing it.